# Assignment Chapter 3 - Machine Learning

## 1. Tentukan Machine Learning

**1. Time Series Forecasting (Peramalan Deret Waktu)**
- Metode prediksi yang digunakan untuk memproyeksikan nilai-nilai masa depan dari suatu variabel berdasarkan data historis yang tersusun secara berurutan berdasarkan waktu. Data time series adalah kumpulan pengamatan atau data yang dikumpulkan pada interval waktu tertentu, misalnya harian, bulanan, atau tahunan. Contoh umum dari data time series adalah harga saham, permintaan listrik, produksi pabrik, dan suhu cuaca.


##2. Sertakan alasan penggunaan/pemilihan machine learning model pada poin 1

- Data Berurutan Secara Waktu: Jika data yang dimiliki berurutan berdasarkan waktu, dan ada pola yang jelas yang muncul secara teratur seiring waktu (seperti tren atau musiman), maka time series forecasting adalah metode yang paling tepat. Dalam situasi ini, data dari waktu sebelumnya dapat memberikan informasi yang sangat berharga tentang nilai-nilai di masa depan.

- Kebutuhan untuk Prediksi Jangka Pendek atau Menengah: Time series forecasting umumnya sangat efektif untuk prediksi jangka pendek atau menengah. Misalnya, jika Anda perlu memprediksi penjualan untuk bulan depan atau permintaan energi di minggu berikutnya, model time series dapat memberikan hasil prediksi yang cukup akurat karena pola musiman atau tren dapat dikenali dalam jangka waktu ini.

- Efisiensi dalam Mengatasi Masalah Musiman atau Siklus: Jika data memiliki sifat musiman, seperti permintaan produk yang meningkat saat liburan, time series forecasting dapat membantu mengidentifikasi dan memperhitungkan siklus musiman ini.




## 3. Eksekusi Coding Machine Learning

In [ ]:
# Import library yang dibutuhkan
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', '{:.2f}'.format)

# Proses ekstraksi data
green_growth_mys = pd.read_csv('GREEN_GROWTH_MYS.csv')

# Tampilkan hasilnya
display(green_growth_mys)

STRUCTURE                              STRUCTURE_ID STRUCTURE_NAME  \
0     DATAFLOW  OECD.ENV.EPI:DSD_GG@DF_GREEN_GROWTH(1.1)   Green Growth   
1     DATAFLOW  OECD.ENV.EPI:DSD_GG@DF_GREEN_GROWTH(1.1)   Green Growth   
2     DATAFLOW  OECD.ENV.EPI:DSD_GG@DF_GREEN_GROWTH(1.1)   Green Growth   
3     DATAFLOW  OECD.ENV.EPI:DSD_GG@DF_GREEN_GROWTH(1.1)   Green Growth   
4     DATAFLOW  OECD.ENV.EPI:DSD_GG@DF_GREEN_GROWTH(1.1)   Green Growth   
...        ...                                       ...            ...   
2531  DATAFLOW  OECD.ENV.EPI:DSD_GG@DF_GREEN_GROWTH(1.1)   Green Growth   
2532  DATAFLOW  OECD.ENV.EPI:DSD_GG@DF_GREEN_GROWTH(1.1)   Green Growth   
2533  DATAFLOW  OECD.ENV.EPI:DSD_GG@DF_GREEN_GROWTH(1.1)   Green Growth   
2534  DATAFLOW  OECD.ENV.EPI:DSD_GG@DF_GREEN_GROWTH(1.1)   Green Growth   
2535  DATAFLOW  OECD.ENV.EPI:DSD_GG@DF_GREEN_GROWTH(1.1)   Green Growth   

     ACTION REF_AREA Reference area FREQ Frequency of observation  \
0         I      MYS       Malaysia    A                   Annual   
1         I      MYS       Malaysia    A                   Annual   
2         I      MYS       Malaysia    A                   Annual   
3         I      MYS       Malaysia    A                   Annual   
4         I      MYS       Malaysia    A                   Annual   
...     ...      ...            ...  ...                      ...   
2531      I      MYS       Malaysia    A                   Annual   
2532      I      MYS       Malaysia    A                   Annual   
2533      I      MYS       Malaysia    A                   Annual   
2534      I      MYS       Malaysia    A                   Annual   
2535      I      MYS       Malaysia    A                   Annual   

            MEASURE    UNIT_MEASURE  \
0            SOCECO              _X   
1            SOCECO              _X   
2            SOCECO              _X   
3            SOCECO              _X   
4            SOCECO              _X   
...             ...             ...   
2531      DRIN_SPOP          PT_POP   
2532  RE_TPES_EBIOM      PT_SUP_NRG   
2533       NRG_PROD         USD_TOE   
2534     ENVRESPROD              _X   
2535     ENVTAX_VEH  PT_TAX_REV_ENV   

                                    Unit of measure ACTIVITY  \
0                                       Unspecified        A   
1                                       Unspecified        A   
2                                       Unspecified        A   
3                                       Unspecified        A   
4                                       Unspecified        A   
...                                             ...      ...   
2531                       Percentage of population       _T   
2532                    Percentage of energy supply       _T   
2533         US dollars per tonne of oil equivalent       _T   
2534                                    Unspecified        A   
2535  Percentage of environment related tax revenue       _T   

                      Economic activity  TIME_PERIOD  OBS_VALUE OBS_STATUS  \
0     Agriculture, forestry and fishing         2017       9.68          A   
1     Agriculture, forestry and fishing         2016       9.68          A   
2     Agriculture, forestry and fishing         2015       9.68          A   
3     Agriculture, forestry and fishing         2014       9.68          A   
4     Agriculture, forestry and fishing         2013       9.68          A   
...                                 ...          ...        ...        ...   
2531             Total - all activities         2008      93.00          A   
2532             Total - all activities         2008       0.93          A   
2533             Total - all activities         2008    7417.82          A   
2534  Agriculture, forestry and fishing         2010       9.68          A   
2535             Total - all activities         2004      99.91          A   

     Observation status  UNIT_MULT Unit multiplier  BASE_PER  TIMELINESS  \
0          Normal value 

- Import Library:

   - pandas: Digunakan untuk memproses dan menganalisis data dalam bentuk tabel (DataFrame). Ini adalah salah satu library paling umum digunakan untuk manipulasi data dalam Python.
- Pengaturan Tampilan:

  - pd.set_option('display.max_columns', None): Ini memastikan bahwa ketika DataFrame ditampilkan, semua kolom akan muncul tanpa terpotong, yang berguna saat bekerja dengan dataset yang memiliki banyak kolom.
  - pd.set_option('display.float_format', '{:.2f}'.format): Mengatur format tampilan angka desimal agar memiliki dua tempat desimal, sehingga data numerik terlihat lebih rapi dan mudah dibaca.
- Ekstraksi Data:
  - green_growth_mys = pd.read_csv('GREEN_GROWTH_MYS.csv'): Membaca file CSV yang berisi data terkait pertumbuhan hijau (green growth) di Malaysia. Data dari file CSV diubah menjadi DataFrame green_growth_mys yang memungkinkan Anda melakukan berbagai operasi analisis dan manipulasi data lebih lanjut.
- Tampilkan Data:
  - display(green_growth_mys): Menampilkan isi DataFrame green_growth_mys. Ini akan menunjukkan data yang telah dimuat dari file CSV dalam bentuk tabel, sehingga Anda dapat melihat isi dan struktur datanya.

In [ ]:
green_growth_mys.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2536 entries, 0 to 2535
Data columns (total 22 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   STRUCTURE                 2536 non-null   object 
 1   STRUCTURE_ID              2536 non-null   object 
 2   STRUCTURE_NAME            2536 non-null   object 
 3   ACTION                    2536 non-null   object 
 4   REF_AREA                  2536 non-null   object 
 5   Reference area            2536 non-null   object 
 6   FREQ                      2536 non-null   object 
 7   Frequency of observation  2536 non-null   object 
 8   MEASURE                   2536 non-null   object 
 9   UNIT_MEASURE              2536 non-null   object 
 10  Unit of measure           2536 non-null   object 
 11  ACTIVITY                  2536 non-null   object 
 12  Economic activity         2536 non-null   object 
 13  TIME_PERIOD               2536 non-null   int64  
 14  OBS_VALU

Code ini berfungsi untuk menampilkan hal seperti kolom, data yang null, dan juga type datanya

In [ ]:
# Definisikan kolom yang dibutuhkan
column = ['Reference area', 'MEASURE', 'Unit of measure', 'Economic activity', 'TIME_PERIOD', 'OBS_VALUE', 'Unit multiplier']

# Ambil kolom yang dibutuhkan saja
green_growth_mys = green_growth_mys.loc[:, column]

# Ubah nama kolom menjadi lowercase dan ubah spasi menjadi underscore
green_growth_mys.columns = green_growth_mys.columns.str.lower()
green_growth_mys.columns = green_growth_mys.columns.str.replace(' ', '_')

# Tampilkan hasilnya
display(green_growth_mys)

- Definisikan kolom yang dibutuhkan
  
  Bagian ini menentukan kolom-kolom spesifik dari dataset yang akan digunakan dalam analisis. Kolom-kolom yang dipilih adalah:
  - Reference area: Mungkin mengacu pada area geografis atau wilayah referensi dari data.
  - MEASURE: Ukuran atau indikator yang digunakan (misalnya, emisi CO2).
  - Unit of measure: Satuan pengukuran (seperti ton, liter, dll.).
  - Economic activity: Aktivitas ekonomi terkait.
  - TIME_PERIOD: Waktu atau periode yang terkait dengan data.
  - OBS_VALUE: Nilai observasi atau data numerik.
  - Unit multiplier: Pengali untuk nilai unit (misalnya ribuan, jutaan, dsb.).
- green_growth_mys = green_growth_mys.loc[:, column]
  - code ini berfungsi mengambil kolom yang telah ditentukan di atas tadi
- str.lower(): Mengubah semua nama kolom menjadi huruf kecil, yang membuatnya konsisten dan lebih mudah diakses dalam kode.
- str.replace(' ', '_'): Mengganti spasi di antara nama kolom dengan garis bawah (_). Ini dilakukan agar nama kolom lebih "kompatibel" dengan sintaks Python, terutama saat menggunakan nama kolom dalam kode.
- display(green_growth_mys)
  - kode ini untuk menampilkan hasilnya

In [ ]:
# Hapus duplikasi
master_measure = green_growth_mys[['economic_activity', 'measure', 'unit_of_measure', 'unit_multiplier']].drop_duplicates()

# Urutkan data
master_measure = master_measure.sort_values(by = ['measure', 'economic_activity'], ignore_index = True)

# Simpan ke excel
master_measure.to_excel('Master Measure.xlsx', index = False, engine = 'openpyxl')

# Tampilkan hasilnya
master_measure

- master_measure = green_growth_mys[['economic_activity', 'measure', 'unit_of_measure', 'unit_multiplier']].drop_duplicates()
   - Bagian ini memilih subset kolom yang spesifik: economic_activity, measure, unit_of_measure, dan unit_multiplier dari dataset utama green_growth_mys.
   - drop_duplicates() digunakan untuk menghapus baris-baris yang memiliki data yang sama di semua kolom yang dipilih. Ini memastikan bahwa tidak ada entri duplikat dalam data yang Anda simpan ke dalam variabel master_measure.
- master_measure = master_measure.sort_values(by = ['measure', 'economic_activity'], ignore_index = True)
   - Data diurutkan berdasarkan kolom measure terlebih dahulu, lalu diikuti oleh economic_activity. Ini memudahkan untuk melihat kategori atau indikator tertentu dan aktivitas ekonomi yang terkait.
   - ignore_index=True memastikan bahwa indeks baru ditetapkan setelah pengurutan, mengabaikan indeks lama.
- master_measure.to_excel('Master Measure.xlsx', index = False, engine = 'openpyxl')
   - Bagian ini menyimpan DataFrame yang telah diproses (master_measure) ke dalam file Excel dengan nama Master Measure.xlsx.
   - index=False berarti indeks tidak akan disertakan dalam file Excel. Ini biasanya diinginkan ketika Anda tidak ingin kolom indeks (yang hanya bilangan bulat yang disediakan oleh pandas) menjadi bagian dari output.
   - engine='openpyxl' digunakan untuk menentukan engine yang akan menulis file Excel.
- master_measure
  - Menampilkan DataFrame master_measure yang telah diproses untuk dilihat langsung hasil pengurutan dan penghapusan duplikasi.



In [ ]:
# Gabungkan kolom
green_growth_mys['measure'] = green_growth_mys['measure'] + ' (' + green_growth_mys['economic_activity'] + ')'

# Tampilkan hasilnya
display(green_growth_mys)

- green_growth_mys['measure'] = green_growth_mys['measure'] + ' (' + green_growth_mys['economic_activity'] + ')'
  - Pada bagian ini, kolom measure dan economic_activity digabungkan menjadi satu.
  - Sintaks + digunakan untuk menggabungkan string dari kolom measure dengan string dalam tanda kurung ( diikuti oleh nilai dalam kolom economic_activity dan ditutup dengan ).
  - Hasil akhirnya adalah kolom measure yang sekarang berisi informasi gabungan dari nama ukuran (measure) dan aktivitas ekonomi terkait, dalam format: measure (economic_activity).
- display(green_growth_mys)
  - Bagian ini menampilkan DataFrame setelah proses penggabungan dilakukan. Anda dapat melihat bahwa kolom measure kini berisi gabungan data dari kolom measure sebelumnya dengan kolom economic_activity.

In [ ]:
# Hapus kolom economic_activity
green_growth_mys = green_growth_mys.drop(columns = 'economic_activity')

# Tampilkan hasilnya
display(green_growth_mys)

reference_area                                         measure  \
0          Malaysia      SOCECO (Agriculture, forestry and fishing)   
1          Malaysia      SOCECO (Agriculture, forestry and fishing)   
2          Malaysia      SOCECO (Agriculture, forestry and fishing)   
3          Malaysia      SOCECO (Agriculture, forestry and fishing)   
4          Malaysia      SOCECO (Agriculture, forestry and fishing)   
...             ...                                             ...   
2531       Malaysia              DRIN_SPOP (Total - all activities)   
2532       Malaysia          RE_TPES_EBIOM (Total - all activities)   
2533       Malaysia               NRG_PROD (Total - all activities)   
2534       Malaysia  ENVRESPROD (Agriculture, forestry and fishing)   
2535       Malaysia             ENVTAX_VEH (Total - all activities)   

                                    unit_of_measure  time_period  obs_value  \
0                                       Unspecified         2017       9.68   
1                                       Unspecified         2016       9.68   
2                                       Unspecified         2015       9.68   
3                                       Unspecified         2014       9.68   
4                                       Unspecified         2013       9.68   
...                                             ...          ...        ...   
2531                       Percentage of population         2008      93.00   
2532                    Percentage of energy supply         2008       0.93   
2533         US dollars per tonne of oil equivalent         2008    7417.82   
2534                                    Unspecified         2010       9.68   
2535  Percentage of environment related tax revenue         2004      99.91   

     unit_multiplier  
0              Units  
1              Units  
2              Units  
3              Units  
4              Units  
...              ...  
2531           Units  
2532           Units  
2533           Units  
2534           Units  
2535           Units  

[2536 rows x 6 columns]

- Kode ini berguna untuk menghapus kolom economy_activity karena kolom nya sudah bergabung dengan measure, jadi hanya perlu menampilkan kolom measure  yang sudah bergabung dengan kolom economy_activity

In [ ]:
# Lakukan pivot table
pvt_green_growth_mys = pd.pivot_table(
    green_growth_mys,
    index = ['reference_area', 'time_period'],
    columns = 'measure',
    values = 'obs_value',
).reset_index()

# Tampilkan hasilnya
display(pvt_green_growth_mys.head(10))

measure reference_area  time_period  ASEW (Total - all activities)  \
0             Malaysia         2000                           9.68   
1             Malaysia         2001                           9.68   
2             Malaysia         2002                           9.68   
3             Malaysia         2003                           9.68   
4             Malaysia         2004                           9.68   
5             Malaysia         2005                           9.68   
6             Malaysia         2006                           9.68   
7             Malaysia         2007                           9.68   
8             Malaysia         2008                           9.68   
9             Malaysia         2009                           9.68   

measure  CO2_AIRTRACAP (Total - all activities)  \
0                                           NaN   
1                                           NaN   
2                                           NaN   
3                                           NaN   
4                                           NaN   
5                                           NaN   
6                                           NaN   
7                                           NaN   
8                                           NaN   
9                                           NaN   

measure  CO2_AIRTRAGDP (Total - all activities)  \
0                                           NaN   
1                                           NaN   
2                                           NaN   
3                                           NaN   
4                                           NaN   
5                                           NaN   
6                                           NaN   
7                                           NaN   
8                                           NaN   
9                                           NaN   

measure  CO2_DBEM (Total - all activities)  \
0                                    83.04   
1                                    97.24   
2                                   107.84   
3                                   105.78   
4                                   113.06   
5                                   118.47   
6                                   119.19   
7                                   135.14   
8                                   141.81   
9                                   145.47   

measure  CO2_DBEMCAP (Total - all activities)  \
0                                        2.88   
1                                        3.29   
2                                        3.55   
3                                        3.40   
4                                        3.55   
5                                        3.64   
6                                        3.58   
7                                        3.97   
8                                        4.08   
9                                        4.10   

measure  CO2_DBPROD (Total - all activities)  \
0                                       5.59   
1                                       4.80   
2                                       4.56   
3                                       4.92   
4                                       4.91   
5                                       4.94   
6                                       5.18   
7                                       4.86   
8                                       4.85   
9                                       4.66   

measure  CO2_PBEM (Total - all activities)  \
0                                   107.53   
1                                   112.56   
2                                   118.22   
3                                   124.10   
4                                   136.98   
5                                   146.84   
6                                   152.48   
7                                   167.67   
8                                   178.99   
9                                   159.72   

measure  CO2_PBEMCAP (

 - index=['reference_area', 'time_period']: Anda menetapkan reference_area (area referensi) dan time_period (periode waktu) sebagai indeks. Ini berarti setiap kombinasi dari area referensi dan periode waktu akan membentuk baris dalam DataFrame hasil pivot.
 - columns='measure': Kolom measure akan diubah menjadi kolom-kolom baru di pivot table. Setiap jenis measure (dengan economic_activity yang telah digabungkan) akan menjadi kolom tersendiri.
 - values='obs_value': Kolom obs_value berisi nilai-nilai yang akan ditempatkan pada persilangan antara baris dan kolom baru yang dihasilkan dari pivot table.
 - reset_index(): Setelah proses pivot, indeks multi-level yang terbentuk dari reference_area dan time_period akan direset menjadi kolom-kolom biasa untuk memudahkan akses.

In [ ]:
# Tampilkan jumlah data
print(f'Jumlah Data : {pvt_green_growth_mys.shape[0]} Baris\n')

# Hitung missing value
data_kosong = pvt_green_growth_mys.isna().sum().to_frame(name = 'Jumlah Missing Value').reset_index()
data_kosong['Persentase NULL %'] = 100 * (data_kosong['Jumlah Missing Value'] /pvt_green_growth_mys.shape[0]).round(4)

# Tampilkan hasilnya
data_kosong = data_kosong.style\
    .map(lambda val : 'background-color: red' if val > 50 else '', subset=['Persentase NULL %'])\
    .format({'Persentase NULL %': '{:.2f}'})

display(data_kosong)

Jumlah Data : 23 Baris



- print(f'Jumlah Data : {pvt_green_growth_mys.shape[0]} Baris\n')
 - pvt_green_growth_mys.shape[0]: Mengambil jumlah baris dalam DataFrame pvt_green_growth_mys.
 - Kode ini akan menampilkan jumlah baris data setelah pivot, untuk mengetahui berapa banyak observasi yang tersedia setelah transformasi.
- data_kosong = pvt_green_growth_mys.isna().sum().to_frame(name = 'Jumlah Missing Value').reset_index()
data_kosong['Persentase NULL %'] = 100 * (data_kosong['Jumlah Missing Value'] /pvt_green_growth_mys.shape[0]).round(4)
 - pvt_green_growth_mys.isna().sum(): Menghitung jumlah missing values (nilai yang kosong atau NaN) di setiap kolom.
 - to_frame(name='Jumlah Missing Value'): Mengubah hasil perhitungan missing values menjadi DataFrame, dengan nama kolom "Jumlah Missing Value".
 - reset_index(): Menyusun ulang indeks sehingga kolom asli dapat menjadi bagian dari DataFrame.
 - Persentase Missing Value: Kolom baru Persentase NULL % dihitung dengan formula (jumlah missing value / total baris) * 100. Ini memberikan informasi berapa persentase nilai kosong di setiap kolom.
- data_kosong = data_kosong.style\
    .map(lambda val : 'background-color: red' if val > 50 else '', subset=['Persentase NULL %'])\
    .format({'Persentase NULL %': '{:.2f}'})
  - map(): Menerapkan fungsi lambda untuk memberikan background berwarna merah pada nilai Persentase NULL % yang lebih dari 50%. Ini menandakan kolom-kolom yang memiliki missing value lebih dari setengah jumlah data sebagai perhatian khusus.
  - format(): Memformat nilai Persentase NULL % agar ditampilkan dengan dua digit desimal.


In [ ]:
# Daftar kolom yang akan dihapus
kolom_yang_dihapus = [
    'SW_NOTTOPERM (Total - all activities)',
    'SW_PERMTONOT (Total - all activities)',
    'SW_PERMTOSEAS (Total - all activities)',
    'SW_PERMWAT (Total - all activities)',
    'SW_SEASTOPERM (Total - all activities)',
    'SW_SEASWAT (Total - all activities)',
    'FOR_IF_EXTENT (Total - all activities)',
    'FOR_IF_PCLOSS2000 (Total - all activities)',
    'FOR_LTMGMT (Total - all activities)',
    'FOR_NAT (Total - all activities)',
    'FOR_STOCK (Total - all activities)',
    'EPRICE_IND (Total - all activities)',
    'EPRICE_RES (Total - all activities)',
    'CO2_AIRTRACAP (Total - all activities)',
    'CO2_AIRTRAGDP (Total - all activities)'
]

# Menghapus kolom dari DataFrame
pvt_green_growth_mys_cleaned = pvt_green_growth_mys.drop(columns=kolom_yang_dihapus)

# Tampilkan jumlah data setelah menghapus kolom
print(f'Jumlah Data setelah menghapus kolom: {pvt_green_growth_mys_cleaned.shape[0]} Baris, {pvt_green_growth_mys_cleaned.shape[1]} Kolom\n')

# Simpan dataset yang telah dibersihkan ke dalam file CSV baru
pvt_green_growth_mys_cleaned.to_csv('cleaned_pivot_green_growth_mys.csv', index=True)

# Tampilkan 5 baris pertama dari data yang telah dibersihkan
print("Contoh data setelah penghapusan kolom:")
print(pvt_green_growth_mys_cleaned.head())


- List kolom_yang_dihapus: Ini adalah daftar kolom yang tidak diperlukan dan akan dihapus dari DataFrame. Kolom ini mungkin mengandung banyak missing values atau tidak relevan untuk analisis lebih lanjut.
- drop(columns=kolom_yang_dihapus): Menghapus kolom-kolom yang ada dalam daftar kolom_yang_dihapus dari DataFrame pvt_green_growth_mys. Hasilnya disimpan dalam DataFrame baru bernama pvt_green_growth_mys_cleaned.
- pvt_green_growth_mys_cleaned.shape: Menampilkan jumlah baris dan kolom dalam DataFrame yang telah dibersihkan setelah penghapusan kolom-kolom yang tidak diperlukan. Ini membantu memeriksa apakah penghapusan kolom berhasil dan seberapa besar pengaruhnya terhadap jumlah kolom.
- to_csv(): Menyimpan DataFrame yang sudah dibersihkan ke dalam file CSV bernama cleaned_pivot_green_growth_mys.csv. Menyimpan data dalam bentuk CSV memungkinkan Anda untuk menggunakan data yang telah dibersihkan untuk analisis atau pemrosesan lebih lanjut.
- head(): Menampilkan 5 baris pertama dari data yang telah dibersihkan untuk memberikan gambaran sekilas tentang hasil akhir setelah penghapusan kolom.

### Code Machine Learning

In [ ]:
# Import pustaka yang diperlukan
import pandas as pd
import warnings
from statsmodels.tsa.arima.model import ARIMA
from sklearn.metrics import mean_absolute_error, mean_squared_error

# Menonaktifkan pesan peringatan yang tidak diinginkan
warnings.filterwarnings("ignore", message="Non-stationary starting autoregressive parameters")
warnings.filterwarnings("ignore", message="Non-invertible starting MA parameters")
warnings.filterwarnings("ignore", message="Maximum Likelihood optimization failed to converge")

# Memuat dataset
data = pd.read_csv('cleaned_pivot_green_growth_mys.csv', index_col='time_period', parse_dates=True)

# Mengisi nilai yang hilang menggunakan metode yang diperbarui
data = data.ffill().bfill()

# Memilih measure yang akan diprediksi
target_measure = 'CO2_DBEM (Total - all activities)'
# Mengambil data time series untuk target
y = data[target_measure]

# Memastikan data memiliki frekuensi yang ditentukan
y = y.asfreq('YS')  # Mengasumsikan data tahunan yang dimulai pada bulan Januari

# Menggunakan data asli tanpa differencing
print("Menggunakan data asli tanpa differencing.")

# Memisahkan data menjadi data training dan testing
train_size = int(len(y) * 0.8)
train, test = y[:train_size], y[train_size:]

# Inisialisasi dan melatih model ARIMA dengan data asli
model = ARIMA(train, order=(1, 1, 1))  # Anda bisa mencoba berbagai kombinasi (p, d, q) sesuai hasil ACF dan PACF
model_fit = model.fit()

# Melakukan prediksi menggunakan data testing
predictions = model_fit.forecast(steps=len(test))

# Mengevaluasi kinerja model
mae = mean_absolute_error(test, predictions)
mse = mean_squared_error(test, predictions)
rmse = mse ** 0.5

# Menampilkan hasil evaluasi
print(f'Mean Absolute Error (MAE): {mae}')
print(f'Mean Squared Error (MSE): {mse}')
print(f'Root Mean Squared Error (RMSE): {rmse}')

# Menampilkan beberapa hasil prediksi pertama beserta data aslinya
results_df = pd.DataFrame({'Actual': test, 'Predicted': predictions}, index=test.index)
print(results_df.head())

# Membuat prediksi masa depan hingga tahun 2030
future_steps = 2030 - data.index[-1].year + 1  # Menghitung jumlah tahun ke depan
future_predictions = model_fit.forecast(steps=future_steps)

# Menghasilkan tanggal masa depan dengan frekuensi yang benar
future_years = pd.date_range(start=data.index[-1] + pd.DateOffset(years=1), periods=future_steps, freq='YS')
future_results = pd.DataFrame({'Year': future_years.year, 'Predicted': future_predictions})

# Menampilkan prediksi hingga tahun 2030
print("\nPrediksi hingga tahun 2030:")
print(future_results)


Menggunakan data asli tanpa differencing.
Mean Absolute Error (MAE): 11.307701930675439
Mean Squared Error (MSE): 191.5279883858896
Root Mean Squared Error (RMSE): 13.839363727638984
             Actual  Predicted
time_period                   
2018-01-01   239.29     232.75
2019-01-01   239.29     240.37
2020-01-01   239.29     247.98
2021-01-01   239.29     255.59
2022-01-01   239.29     263.21

Prediksi hingga tahun 2030:
            Year  Predicted
2018-01-01  2023     232.75
2019-01-01  2024     240.37
2020-01-01  2025     247.98
2021-01-01  2026     255.59
2022-01-01  2027     263.21
2023-01-01  2028     270.82
2024-01-01  2029     278.43
2025-01-01  2030     286.05
2026-01-01  2031     293.66


- Impor Pustaka yang Diperlukan:

  - pandas digunakan untuk manipulasi data.
  - warnings digunakan untuk menonaktifkan peringatan yang tidak diperlukan dari model ARIMA.
  - ARIMA dari statsmodels digunakan untuk membuat model ARIMA.
  - mean_absolute_error dan mean_squared_error dari sklearn.metrics digunakan untuk mengevaluasi kinerja model.
- Menonaktifkan Peringatan:
  - Peringatan yang berhubungan dengan parameter model ARIMA yang tidak stasioner atau gagal dikonvergensi dinonaktifkan untuk membuat output lebih bersih.
- Memuat Dataset:
  - Dataset yang sudah dibersihkan dan disimpan dalam file 'cleaned_pivot_green_growth_mys.csv' dimuat. Data diindeks berdasarkan time_period yang diparse sebagai format tanggal.
- Mengisi Nilai yang Hilang:
 - Metode ffill() dan bfill() digunakan untuk mengisi missing value. Ini memastikan tidak ada nilai yang hilang dalam data time series, dengan menggunakan pengisian maju dan mundur.
- Memilih Kolom Target:
 - Variabel target yang dipilih untuk prediksi adalah 'CO2_DBEM (Total - all activities)', yaitu jumlah emisi CO2 dari seluruh aktivitas.
- Mengatur Frekuensi Data:
  - Frekuensi data diatur menjadi tahunan ('YS'), yang artinya diasumsikan bahwa setiap titik data merupakan data tahunan yang dimulai dari bulan Januari.
- Membagi Data menjadi Training dan Testing:
  - Data time series dibagi menjadi 80% data untuk training dan 20% untuk testing.
- Model ARIMA:

  - Model ARIMA diinisialisasi dengan parameter (1, 1, 1), di mana:
    - p=1 adalah lag untuk komponen autoregressive (AR).
    - d=1 adalah jumlah differencing yang dilakukan agar data menjadi stasioner.
    - q=1 adalah lag untuk komponen moving average (MA).
  - Model dilatih menggunakan data training.
- Prediksi dan Evaluasi:
  - Setelah model dilatih, dilakukan prediksi pada data testing.
  - Evaluasi model dilakukan menggunakan tiga metrik:
    - MAE (Mean Absolute Error): Rata-rata absolut dari selisih antara nilai aktual dan nilai prediksi.
    - MSE (Mean Squared Error): Rata-rata dari kuadrat selisih antara nilai aktual dan prediksi.
    - RMSE (Root Mean Squared Error): Akar kuadrat dari MSE, memberikan gambaran seberapa jauh prediksi dari nilai aktual dalam satuan asli.
- Menampilkan Hasil Prediksi:
  - Beberapa prediksi pertama dari model dibandingkan dengan data aktual.
- Prediksi Masa Depan Hingga Tahun 2030:
  - Model digunakan untuk memprediksi nilai CO2 hingga tahun 2030.
  - Tanggal masa depan dihasilkan dengan frekuensi tahunan, dan prediksi dibuat untuk tahun-tahun tersebut.

- Interpretasi Hasil :

1. Menggunakan Data Asli tanpa Differencing:
   Differencing biasanya digunakan untuk membuat data time series menjadi stasioner, tetapi di sini model ARIMA dilatih tanpa differencing. Ini berarti model menggunakan data aslinya tanpa mengubah tren atau pola data.
2. Evaluasi Kinerja Model:
   - Mean Absolute Error (MAE): 11.31: Rata-rata dari selisih absolut antara nilai aktual dan prediksi adalah sekitar 11.31. MAE memberikan gambaran seberapa jauh prediksi dari nilai aktual dalam satuan asli.
   - Mean Squared Error (MSE): 191.53: MSE menunjukkan rata-rata kuadrat dari kesalahan prediksi, yang memberikan bobot lebih besar untuk kesalahan yang lebih besar. Nilai MSE ini cukup besar, menandakan ada beberapa deviasi yang signifikan antara nilai aktual dan prediksi.
   - Root Mean Squared Error (RMSE): 13.84: RMSE adalah akar kuadrat dari MSE. Ini memberikan gambaran umum mengenai kesalahan prediksi dalam satuan asli (CO2 emissions). Nilai RMSE yang lebih tinggi dari MAE menunjukkan adanya kesalahan besar pada beberapa titik data.
3. Hasil Prediksi pada Data Testing:
   - Pada tabel hasil prediksi, "Actual" adalah data aktual yang diambil dari dataset, dan "Predicted" adalah hasil prediksi dari model ARIMA.
   - Kita bisa melihat prediksi cukup dekat dengan nilai aktual, meskipun ada sedikit penyimpangan:
   - Pada tahun 2018, prediksi adalah 232.75, sementara nilai aktual adalah 239.29, sehingga ada perbedaan kecil.
   - Prediksi terus meningkat dari tahun ke tahun, sesuai dengan tren data, tetapi masih ada penyimpangan kecil dibandingkan dengan data aktual.
4. Prediksi Masa Depan Hingga Tahun 2030:
   - Prediksi untuk Tahun 2023-2030: Model ARIMA telah digunakan untuk membuat prediksi emisi CO2 hingga tahun 2030.
   - Prediksi menunjukkan tren kenaikan bertahap:
   - Pada tahun 2023, prediksi emisi CO2 adalah 232.75.
   - Emisi CO2 diperkirakan akan terus meningkat hingga 286.05 pada tahun 2030.

Hal ini mengindikasikan bahwa model memperkirakan adanya peningkatan konstan pada emisi CO2 dari tahun ke tahun.

## 4. Lakukan Cross-Validation Basic dari machine learning model yang telah dibuat sebelumnya

In [ ]:
# Import pustaka yang diperlukan
from statsmodels.tsa.arima.model import ARIMA
from sklearn.metrics import mean_absolute_error, mean_squared_error
import numpy as np

# Fungsi untuk melakukan cross-validation
def time_series_cross_validation(data, target_measure, p, d, q, n_splits=5):
    y = data[target_measure]
    y = y.asfreq('YS')  # Mengatur frekuensi data

    # Panjang setiap fold dalam cross-validation
    fold_size = int(len(y) / n_splits)

    # Menyimpan hasil evaluasi
    mae_scores = []
    mse_scores = []
    rmse_scores = []

    for i in range(1, n_splits):
        # Tentukan train dan test set berdasarkan split
        train = y[:fold_size * i]
        test = y[fold_size * i: fold_size * (i + 1)]

        # Melatih model ARIMA pada data train
        model = ARIMA(train, order=(p, d, q))
        model_fit = model.fit()

        # Prediksi pada data test
        predictions = model_fit.forecast(steps=len(test))

        # Evaluasi prediksi dengan MAE, MSE, dan RMSE
        mae = mean_absolute_error(test, predictions)
        mse = mean_squared_error(test, predictions)
        rmse = mse ** 0.5

        # Simpan hasil evaluasi
        mae_scores.append(mae)
        mse_scores.append(mse)
        rmse_scores.append(rmse)

    # Rata-rata hasil evaluasi
    mean_mae = np.mean(mae_scores)
    mean_mse = np.mean(mse_scores)
    mean_rmse = np.mean(rmse_scores)

    # Menampilkan hasil cross-validation secara keseluruhan
    print("\nHasil Cross-Validation:")
    print(f"Rata-rata MAE: {mean_mae}")
    print(f"Rata-rata MSE: {mean_mse}")
    print(f"Rata-rata RMSE: {mean_rmse}")

# Memanggil fungsi cross-validation
time_series_cross_validation(data, target_measure='CO2_DBEM (Total - all activities)', p=1, d=1, q=1, n_splits=5)

/usr/local/lib/python3.10/dist-packages/statsmodels/tsa/statespace/sarimax.py:866: UserWarning: Too few observations to estimate starting parameters for ARMA and trend. All parameters except for variances will be set to zeros.
  warn('Too few observations to estimate starting parameters%s.'



Hasil Cross-Validation:
Rata-rata MAE: 17.593190928381738
Rata-rata MSE: 437.74762922474935
Rata-rata RMSE: 20.18367650915617


## 5. Jelaskan Hasil dari Cross-Validation Basic yang telah dihasilkan sebelumnya pada poin 4

### 1. **Mengimpor library yang diperlukan**
- **Warnings** digunakan untuk menonaktifkan beberapa peringatan yang tidak diperlukan selama pelatihan model ARIMA.
- **Pandas** digunakan untuk pengelolaan dan manipulasi data time series.
- **ARIMA** dari **statsmodels.tsa.arima.model** digunakan sebagai model prediksi berbasis time series.

*   **Mean_absolute_error** dan **mean_squared_error dari sklearn.metrics** digunakan untuk mengevaluasi performa prediksi model.
*  **Numpy** digunakan untuk menghitung nilai rata-rata dari hasil evaluasi di beberapa fold.


## 2. **Menyiapkan Model Arima**

*   Model ARIMA dibangun dengan parameter **p=1, d=1, q=1** untuk setiap fold dalam cross-validation. Model ini dilatih pada data time series yang dipartisi sesuai skema cross-validation.
*   Target yang dianalisis adalah **CO2_DBEM (Total - all activities)**, dengan frekuensi data tahunan yang diatur menggunakan **y.asfreq('YS')**.



## 3. **Melakukan Cross-Validation dengan 5 fold**
Fungsi time_series_cross_validation menggunakan cross-validation dengan 5 fold (n_splits=5). Pada setiap fold, data dibagi menjadi set pelatihan dan pengujian, dan model ARIMA dilatih pada set pelatihan sebelum memprediksi nilai pada set pengujian.

## 4. **Melakukan prediksi menggunakan cross-validation**
Setelah mengumpulkan hasil evaluasi dari semua fold, nilai rata-rata performa dihitung sebagai berikut:

**1. Rata-rata MAE (Mean Absolute Error): 17.59**

- MAE adalah ukuran rata-rata dari perbedaan absolut antara nilai aktual dan nilai prediksi. Nilai 17.59 menunjukkan bahwa rata-rata kesalahan prediksi model ARIMA adalah sekitar 17.59 unit dari target yang sebenarnya. Semakin rendah nilai MAE, semakin baik performa model, karena itu berarti prediksinya semakin dekat dengan nilai sebenarnya.

**2. Rata-rata MSE (Mean Squared Error): 437.75**
- RMSE mengukur rata-rata kuadrat perbedaan antara nilai prediksi dan nilai aktual. Nilai yang tinggi, seperti 437.75, menunjukkan adanya beberapa prediksi yang meleset cukup jauh dari nilai sebenarnya, karena MSE memperbesar kesalahan besar dengan menguadratkannya. Ini bisa menjadi indikator bahwa model kadang-kadang menghasilkan prediksi yang melenceng signifikan dari nilai asli.


**3. Rata-rata RMSE (Root Mean Squared Error): 20.18**
- RMSE adalah akar kuadrat dari MSE dan memberikan ukuran kesalahan dalam satuan yang sama dengan data asli. Dengan nilai 20.18, ini menunjukkan bahwa secara rata-rata, kesalahan prediksi model adalah sekitar 20.18 unit dari nilai sebenarnya. Sama seperti MAE, semakin rendah nilai RMSE, semakin baik kualitas prediksi model.

Secara umum, Model ARIMA yang digunakan menghasilkan rata-rata kesalahan prediksi sebesar 17.59 unit (MAE) dan 20.18 unit (RMSE), dengan nilai MSE 437.75 yang mengindikasikan adanya beberapa prediksi yang cukup jauh dari nilai aktual. Meskipun nilai MAE dan RMSE cukup dekat, yang menunjukkan bahwa kesalahan besar tidak terlalu ekstrem, hasil ini menunjukkan bahwa model masih memiliki tingkat kesalahan yang signifikan dalam memprediksi data.

## 6.Terapkan leaming curves dari machine leaming model yang telah dibuat sebelumnya serta membuat visualisasi hasil dari leaming curves yang dihasilkan.

In [ ]:
import numpy as np
import plotly.graph_objects as go

# Fungsi untuk menghitung error metrik
def calculate_metrics(y_true, y_pred):
    mae = mean_absolute_error(y_true, y_pred)
    mse = mean_squared_error(y_true, y_pred)
    rmse = np.sqrt(mse)
    return mae, mse, rmse

# Menyimpan metrik kinerja untuk tiap ukuran data pelatihan
train_sizes = np.linspace(0.1, 1.0, 10)  # Mengambil proporsi data pelatihan dari 10% hingga 100%
train_mae, train_mse, train_rmse = [], [], []
test_mae, test_mse, test_rmse = [], [], []

# Looping melalui ukuran data pelatihan yang berbeda
for train_size in train_sizes:
    train_cut = int(len(train) * train_size)

    # Menggunakan subset dari data pelatihan
    train_subset = train[:train_cut]

    # Inisialisasi dan melatih model ARIMA pada subset data pelatihan
    model = ARIMA(train_subset, order=(1, 1, 1))
    model_fit = model.fit()

    # Prediksi pada data pelatihan
    train_predictions = model_fit.predict(start=train_subset.index[0], end=train_subset.index[-1])

    # Menghitung metrik pada data pelatihan
    mae_train, mse_train, rmse_train = calculate_metrics(train_subset, train_predictions)
    train_mae.append(mae_train)
    train_mse.append(mse_train)
    train_rmse.append(rmse_train)

    # Prediksi pada data uji (test data)
    test_predictions = model_fit.forecast(steps=len(test))

    # Menghitung metrik pada data uji
    mae_test, mse_test, rmse_test = calculate_metrics(test, test_predictions)
    test_mae.append(mae_test)
    test_mse.append(mse_test)
    test_rmse.append(rmse_test)

# Membuat visualisasi interaktif menggunakan Plotly

# MAE Plot
fig_mae = go.Figure()
fig_mae.add_trace(go.Scatter(x=train_sizes, y=train_mae, mode='lines+markers', name='Train MAE', line=dict(color='blue')))
fig_mae.add_trace(go.Scatter(x=train_sizes, y=test_mae, mode='lines+markers', name='Test MAE', line=dict(color='orange')))
fig_mae.update_layout(title='Learning Curve - MAE', xaxis_title='Training Size', yaxis_title='MAE')

# MSE Plot
fig_mse = go.Figure()
fig_mse.add_trace(go.Scatter(x=train_sizes, y=train_mse, mode='lines+markers', name='Train MSE', line=dict(color='blue')))
fig_mse.add_trace(go.Scatter(x=train_sizes, y=test_mse, mode='lines+markers', name='Test MSE', line=dict(color='orange')))
fig_mse.update_layout(title='Learning Curve - MSE', xaxis_title='Training Size', yaxis_title='MSE')

# RMSE Plot
fig_rmse = go.Figure()
fig_rmse.add_trace(go.Scatter(x=train_sizes, y=train_rmse, mode='lines+markers', name='Train RMSE', line=dict(color='blue')))
fig_rmse.add_trace(go.Scatter(x=train_sizes, y=test_rmse, mode='lines+markers', name='Test RMSE', line=dict(color='orange')))
fig_rmse.update_layout(title='Learning Curve - RMSE', xaxis_title='Training Size', yaxis_title='RMSE')

# Menampilkan plot interaktif
fig_mae.show()
fig_mse.show()
fig_rmse.show()

/usr/local/lib/python3.10/dist-packages/statsmodels/tsa/statespace/sarimax.py:866: UserWarning: Too few observations to estimate starting parameters for ARMA and trend. All parameters except for variances will be set to zeros.
  warn('Too few observations to estimate starting parameters%s.'
/usr/local/lib/python3.10/dist-packages/statsmodels/tsa/statespace/mlemodel.py:1235: RuntimeWarning: invalid value encountered in divide
  np.inner(score_obs, score_obs) /
/usr/local/lib/python3.10/dist-packages/statsmodels/tsa/statespace/sarimax.py:866: UserWarning: Too few observations to estimate starting parameters for ARMA and trend. All parameters except for variances will be set to zeros.
  warn('Too few observations to estimate starting parameters%s.'


## 7.Jelaskan hasil dari visualisasi learning curves yang telah dihasilkan sebelumnya pada poin 6.

Kode pada poin 6 bertujuan untuk membuat learning curves dari model ARIMA dengan menggunakan berbagai ukuran data pelatihan (dari 10% hingga 100%). Kode ini mencakup langkah-langkah berikut:

**Fungsi calculate_metrics:** Menghitung tiga metrik error—MAE (Mean Absolute Error), MSE (Mean Squared Error), dan RMSE (Root Mean Squared Error)—antara prediksi dan data aktual.

**Menentukan ukuran data pelatihan (train_sizes):** Membagi data pelatihan menjadi beberapa subset dengan proporsi yang berbeda-beda, mulai dari 10% hingga 100% dari total data pelatihan.

**Melatih model ARIMA:** Pada tiap iterasi, model ARIMA dilatih pada subset data pelatihan yang berbeda, dan prediksi dilakukan pada data pelatihan dan data uji.

**Menghitung metrik error:** Setelah melakukan prediksi pada data pelatihan dan uji, MAE, MSE, dan RMSE dihitung dan disimpan untuk setiap subset data pelatihan.

**Visualisasi hasil dengan Plotly:** Membuat tiga plot interaktif yang masing-masing menunjukkan MAE, MSE, dan RMSE terhadap ukuran data pelatihan. Dengan Plotly, pengguna dapat melihat perubahan performa model saat ukuran data pelatihan meningkat.

## 8. Perbaiki akurasi machine learning modul yang telah dibuat menggunakan hyperparameter tuning dan juga bootstrapping, jika ada. Jelaskan hasil dari akurasi model (apakah model yang dibuat sudah baik atau belum) yang dibuat beserta prosesnya

In [ ]:
# hyperparameter tuning
from statsmodels.tsa.arima.model import ARIMA
from sklearn.model_selection import ParameterGrid

# Definisikan grid dari hyperparameter ARIMA
p = range(0, 3)
d = range(0, 2)
q = range(0, 3)
param_grid = ParameterGrid({'p': p, 'd': d, 'q': q})

best_params = None
best_rmse = float('inf')

# Loop melalui semua kombinasi hyperparameter
for params in param_grid:
    try:
        model = ARIMA(train, order=(params['p'], params['d'], params['q']))
        model_fit = model.fit()

        test_predictions = model_fit.forecast(steps=len(test))
        _, _, rmse_test = calculate_metrics(test, test_predictions)

        # Simpan hyperparameter dengan RMSE terbaik
        if rmse_test < best_rmse:
            best_rmse = rmse_test
            best_params = params

    except:
        continue

print(f'Best Hyperparameters: {best_params}')
print(f'Best RMSE: {best_rmse}')

Best Hyperparameters: {'d': 1, 'p': 2, 'q': 1}
Best RMSE: 7.29717585970053


Best Hyperparameters: Kombinasi nilai-nilai p=2, d=1, q=1 memberikan performa terbaik dalam memprediksi data uji, berdasarkan nilai RMSE yang diukur.
- p = 2: Model ARIMA mempertimbangkan hingga 2 lag dari komponen autoregressive.
- d = 1: Model menggunakan satu tingkat differencing untuk mengatasi non-stasioneritas.
- q = 1: Model mempertimbangkan 1 lag dari komponen moving average.
Best RMSE: Nilai RMSE sebesar 7.297 menunjukkan seberapa besar kesalahan prediksi model terhadap data uji.

Semakin rendah RMSE, semakin baik model dalam melakukan prediksi. Nilai ini dapat dibandingkan dengan model lain atau baseline untuk menilai apakah performanya sudah cukup baik.

Kesimpulan:
- Hyperparameter tuning berhasil menemukan kombinasi yang optimal dengan nilai RMSE yang rendah, sehingga model ARIMA ini lebih akurat dalam memprediksi data time series dibandingkan dengan kombinasi lain dari parameter yang diujicoba.

In [ ]:
import numpy as np
import plotly.graph_objects as go
from sklearn.utils import resample
from sklearn.metrics import mean_absolute_error, mean_squared_error
from statsmodels.tsa.arima.model import ARIMA

# Fungsi untuk menghitung error metrik
def calculate_metrics(y_true, y_pred):
    mae = mean_absolute_error(y_true, y_pred)
    mse = mean_squared_error(y_true, y_pred)
    rmse = np.sqrt(mse)
    return mae, mse, rmse

# Fungsi untuk melakukan bootstrapping dan melatih model ARIMA
def bootstrap_arima(train, test, train_size, n_bootstrap=100):
    boot_train_mae, boot_train_mse, boot_train_rmse = [], [], []
    boot_test_mae, boot_test_mse, boot_test_rmse = [], [], []

    for _ in range(n_bootstrap):
        # Melakukan bootstrapping pada data pelatihan
        train_cut = int(len(train) * train_size)
        train_subset = resample(train[:train_cut], replace=True)

        # Inisialisasi dan melatih model ARIMA pada subset data pelatihan
        model = ARIMA(train_subset, order=(1, 1, 1))
        model_fit = model.fit()

        # Prediksi pada data pelatihan
        train_predictions = model_fit.predict(start=train_subset.index[0], end=train_subset.index[-1])

        # Menghitung metrik pada data pelatihan
        mae_train, mse_train, rmse_train = calculate_metrics(train_subset, train_predictions)
        boot_train_mae.append(mae_train)
        boot_train_mse.append(mse_train)
        boot_train_rmse.append(rmse_train)

        # Prediksi pada data uji (test data)
        test_predictions = model_fit.forecast(steps=len(test))

        # Menghitung metrik pada data uji
        mae_test, mse_test, rmse_test = calculate_metrics(test, test_predictions)
        boot_test_mae.append(mae_test)
        boot_test_mse.append(mse_test)
        boot_test_rmse.append(rmse_test)

    # Mengembalikan rata-rata metrik kinerja dari semua iterasi bootstrap
    return (np.mean(boot_train_mae), np.mean(boot_train_mse), np.mean(boot_train_rmse),
            np.mean(boot_test_mae), np.mean(boot_test_mse), np.mean(boot_test_rmse))

# Menyimpan metrik kinerja untuk tiap ukuran data pelatihan
train_sizes = np.linspace(0.1, 1.0, 10)  # Mengambil proporsi data pelatihan dari 10% hingga 100%
train_mae, train_mse, train_rmse = [], [], []
test_mae, test_mse, test_rmse = [], [], []

# MAE Plot
fig_mae = go.Figure()
fig_mae.add_trace(go.Scatter(x=train_sizes, y=train_mae, mode='lines+markers', name='Train MAE', line=dict(color='blue')))
fig_mae.add_trace(go.Scatter(x=train_sizes, y=test_mae, mode='lines+markers', name='Test MAE', line=dict(color='orange')))
fig_mae.update_layout(title='Learning Curve - MAE (with Bootstrapping)', xaxis_title='Training Size', yaxis_title='MAE')

# MSE Plot
fig_mse = go.Figure()
fig_mse.add_trace(go.Scatter(x=train_sizes, y=train_mse, mode='lines+markers', name='Train MSE', line=dict(color='blue')))
fig_mse.add_trace(go.Scatter(x=train_sizes, y=test_mse, mode='lines+markers', name='Test MSE', line=dict(color='orange')))
fig_mse.update_layout(title='Learning Curve - MSE (with Bootstrapping)', xaxis_title='Training Size', yaxis_title='MSE')

# RMSE Plot
fig_rmse = go.Figure()
fig_rmse.add_trace(go.Scatter(x=train_sizes, y=train_rmse, mode='lines+markers', name='Train RMSE', line=dict(color='blue')))
fig_rmse.add_trace(go.Scatter(x=train_sizes, y=test_rmse, mode='lines+markers', name='Test RMSE', line=dict(color='orange')))
fig_rmse.update_layout(title='Learning Curve - RMSE (with Bootstrapping)', xaxis_title='Training Size', yaxis_title='RMSE')

# Menampilkan plot interaktif
fig_mae.show()
fig_mse.show()
fig_rmse.show()

In [ ]:
import numpy as np
import pandas as pd
import plotly.graph_objects as go
from sklearn.utils import resample
from sklearn.metrics import mean_absolute_error, mean_squared_error
from statsmodels.tsa.arima.model import ARIMA
from sklearn.model_selection import train_test_split

# Fungsi untuk menghitung error metrik
def calculate_metrics(y_true, y_pred):
    mae = mean_absolute_error(y_true, y_pred)
    mse = mean_squared_error(y_true, y_pred)
    rmse = np.sqrt(mse)
    return mae, mse, rmse

# Fungsi untuk melakukan bootstrapping dan melatih model ARIMA
def bootstrap_arima(train, test, train_size, n_bootstrap=100):
    boot_train_mae, boot_train_mse, boot_train_rmse = [], [], []
    boot_test_mae, boot_test_mse, boot_test_rmse = [], [], []

    # Check the cut-off for training data
    train_cut = int(len(train) * train_size)

    if train_cut < 2:  # Ensure there are enough observations
        print(f"Not enough training data for train size: {train_size:.2f}")
        return np.nan, np.nan, np.nan, np.nan, np.nan, np.nan

    for _ in range(n_bootstrap):
        # Melakukan bootstrapping pada data pelatihan
        train_subset = resample(train.iloc[:train_cut, 0], replace=True)  # Use only the first column

        # Inisialisasi dan melatih model ARIMA pada subset data pelatihan
        if len(train_subset) < 2:  # Ensure enough data for model fitting
            continue  # Skip this iteration if not enough data

        model = ARIMA(train_subset, order=(1, 1, 1))

        try:
            model_fit = model.fit()
        except Exception as e:
            print(f"Model fitting failed: {e}")
            continue  # Skip this iteration if fitting fails

        # Prediksi pada data pelatihan
        train_predictions = model_fit.predict(start=0, end=len(train_subset) - 1)

        # Menghitung metrik pada data pelatihan
        mae_train, mse_train, rmse_train = calculate_metrics(train_subset, train_predictions)
        boot_train_mae.append(mae_train)
        boot_train_mse.append(mse_train)
        boot_train_rmse.append(rmse_train)

        # Prediksi pada data uji (test data)
        test_predictions = model_fit.forecast(steps=len(test))

        # Menghitung metrik pada data uji
        if len(test) > 0:  # Ensure test data is not empty
            mae_test, mse_test, rmse_test = calculate_metrics(test.iloc[:, 0], test_predictions)  # Use only the first column
            boot_test_mae.append(mae_test)
            boot_test_mse.append(mse_test)
            boot_test_rmse.append(rmse_test)

    # Mengembalikan rata-rata metrik kinerja dari semua iterasi bootstrap
    return (np.mean(boot_train_mae), np.mean(boot_train_mse), np.mean(boot_train_rmse),
            np.mean(boot_test_mae), np.mean(boot_test_mse), np.mean(boot_test_rmse))

# Menyimpan metrik kinerja untuk tiap ukuran data pelatihan
train_sizes = np.linspace(0.1, 1.0, 10)  # Mengambil proporsi data pelatihan dari 10% hingga 100%
train_mae, train_mse, train_rmse = [], [], []
test_mae, test_mse, test_rmse = [], [], []

# Assuming `green_growth_mys` is your full dataset (ensure it is defined)
train, test = train_test_split(green_growth_mys, test_size=0.2, random_state=42)

# Check and convert data types to numeric, if necessary
def check_and_convert_to_numeric(df):
    for column in df.columns:
        df[column] = pd.to_numeric(df[column], errors='coerce')
    return df.dropna()  # Drop rows with NaN values

train = check_and_convert_to_numeric(train)
test = check_and_convert_to_numeric(test)

# Loop untuk tiap ukuran data pelatihan
for train_size in train_sizes:
    mae_train, mse_train, rmse_train, mae_test, mse_test, rmse_test = bootstrap_arima(train, test, train_size)
    train_mae.append(mae_train)
    train_mse.append(mse_train)
    train_rmse.append(rmse_train)

# MAE Plot
fig_mae = go.Figure()
fig_mae.add_trace(go.Scatter(x=train_sizes, y=train_mae, mode='lines+markers', name='Train MAE', line=dict(color='blue')))
fig_mae.add_trace(go.Scatter(x=train_sizes, y=test_mae, mode='lines+markers', name='Test MAE', line=dict(color='orange')))
fig_mae.update_layout(title='Learning Curve - MAE (with Bootstrapping)', xaxis_title='Training Size', yaxis_title='MAE')

# MSE Plot
fig_mse = go.Figure()
fig_mse.add_trace(go.Scatter(x=train_sizes, y=train_mse, mode='lines+markers', name='Train MSE', line=dict(color='blue')))
fig_mse.add_trace(go.Scatter(x=train_sizes, y=test_mse, mode='lines+markers', name='Test MSE', line=dict(color='orange')))
fig_mse.update_layout(title='Learning Curve - MSE (with Bootstrapping)', xaxis_title='Training Size', yaxis_title='MSE')

# RMSE Plot
fig_rmse = go.Figure()
fig_rmse.add_trace(go.Scatter(x=train_sizes, y=train_rmse, mode='lines+markers', name='Train RMSE', line=dict(color='blue')))
fig_rmse.add_trace(go.Scatter(x=train_sizes, y=test_rmse, mode='lines+markers', name='Test RMSE', line=dict(color='orange')))
fig_rmse.update_layout(title='Learning Curve - RMSE (with Bootstrapping)', xaxis_title='Training Size', yaxis_title='RMSE')

# Menampilkan plot interaktif
fig_mae.show()
fig_mse.show()
fig_rmse.show()


Not enough training data for train size: 0.10
Not enough training data for train size: 0.20
Not enough training data for train size: 0.30
Not enough training data for train size: 0.40
Not enough training data for train size: 0.50
Not enough training data for train size: 0.60
Not enough training data for train size: 0.70
Not enough training data for train size: 0.80
Not enough training data for train size: 0.90
Not enough training data for train size: 1.00


Interpretasi Hasil:

- Jika akurasi (MAE, MSE, RMSE) pada data uji terus membaik atau mendekati akurasi pada data pelatihan saat ukuran data meningkat, model bisa dianggap baik.

- Jika terdapat perbedaan besar antara performa pada data pelatihan dan data uji, ini menunjukkan adanya overfitting atau underfitting.

- Bootstrapping membantu mengurangi variasi hasil dari model yang dilatih pada subset data yang berbeda.

Saat mencoba meningkatkan model ARIMA ini dengan bootstrapping, tujuannya adalah untuk melihat seberapa stabil modelnya, serta mencoba menangani fluktuasi performa di data yang lebih kecil. Jadi, saya menggunakan bootstrapping buat melatih ARIMA berkali-kali di subset data yang dipilih secara acak dengan penggantian. Ini diulang 100 kali, dan setiap kali menghitung error metriknya seperti MAE, MSE, dan RMSE.

## 9. Gunakan metric evaluation untuk mengukur model yang telah dibuat sebelumnya

In [ ]:
from sklearn.metrics import mean_absolute_error, mean_squared_error

# Melakukan prediksi menggunakan model ARIMA pada data test
predictions = model_fit.forecast(steps=len(test))

# Menghitung metrik evaluasi untuk prediksi ARIMA
mae = mean_absolute_error(test, predictions)
mse = mean_squared_error(test, predictions)
rmse = np.sqrt(mse)

# Cetak hasil metrik
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"Root Mean Squared Error (RMSE): {rmse:.2f}")


Interpretasi MAE, MSE, dan RMSE:

**Mean Absolute Error (MAE):**
MAE adalah rata-rata perbedaan absolut antara nilai prediksi dan nilai sebenarnya pada data uji. Nilai MAE sebesar 34.82 menunjukkan bahwa, rata-rata, prediksi model ARIMA meleset sekitar 34.82 unit dari nilai aktual. Metrik ini mudah diinterpretasikan karena berada dalam satuan yang sama dengan data asli. Semakin rendah nilai MAE, semakin baik model tersebut.

**Mean Squared Error (MSE):**
MSE mengukur rata-rata kuadrat dari kesalahan antara prediksi dan nilai sebenarnya. Nilai MSE sebesar 1264.65 menyoroti bahwa ada beberapa prediksi yang mengalami kesalahan cukup besar, karena kuadrat dari perbedaan memberikan bobot lebih besar pada kesalahan yang ekstrem. Nilai yang lebih rendah pada MSE menunjukkan bahwa rata-rata kesalahan kuadrat lebih kecil, yang berarti model lebih akurat.

**Root Mean Squared Error (RMSE):**
RMSE adalah akar kuadrat dari MSE dan mengembalikan nilai kesalahan ke dalam satuan yang sama dengan data asli. Nilai RMSE sebesar 35.56 menunjukkan bahwa, secara rata-rata, kesalahan prediksi model ARIMA adalah sekitar 35.56 unit. Seperti halnya MAE, semakin kecil nilai RMSE, semakin baik kinerja model.

**Kesimpulan:**
Nilai MAE dan RMSE yang cukup besar (34.82 dan 35.56) menunjukkan bahwa prediksi model ARIMA ini memiliki tingkat kesalahan yang signifikan. Kesalahan ini dapat dianggap besar, terutama jika data asli memiliki skala yang lebih kecil. Jika data aktual berada dalam rentang nilai yang lebih tinggi, kesalahan ini mungkin dapat diterima. Namun, jika rentang data relatif kecil, maka model perlu diperbaiki atau dioptimalkan lebih lanjut. Upaya perbaikan bisa meliputi penyesuaian parameter model atau mencoba model yang berbeda untuk meningkatkan akurasi prediksi.

## 10. Berikan interpretasi dari hasil metric evaluation serta alasan pemilihan metric yang dipilih pada poin 9

Berikut adalah interpretasi dari hasil metric evaluation pada model yang telah dibuat:

- Accuracy: Mengukur seberapa sering prediksi benar dari total prediksi. Jika accuracy tinggi (misalnya >0.9), maka model mampu memprediksi dengan baik untuk data yang seimbang.
- Precision: Penting jika kita ingin meminimalkan false positives. Jika precision tinggi, berarti sedikit kesalahan dalam mengklasifikasikan kelas positif.
- Recall: Penting jika kita ingin mendeteksi sebanyak mungkin kelas positif (false negatives harus sedikit). Recall tinggi menunjukkan model efektif dalam menangkap semua kasus positif.
- F1-Score: Rata-rata harmonis antara precision dan recall. Dipilih ketika distribusi kelas tidak seimbang, karena memberikan keseimbangan antara false positives dan false negatives.

Alasan Pemilihan Metrik:
- Accuracy baik digunakan jika kelas seimbang, namun tidak cukup jika data memiliki ketidakseimbangan kelas.
- Precision dan Recall memberikan pandangan lebih dalam terhadap kinerja model di kelas minoritas, sehingga lebih sesuai jika ada perbedaan signifikan antara distribusi kelas.
- F1-Score sangat penting untuk data yang memiliki distribusi kelas tidak seimbang, karena memastikan bahwa model baik dalam menemukan kelas minoritas tanpa mengorbankan akurasi keseluruhan.

Kombinasi dari metrik-metrik ini memberikan gambaran lebih lengkap mengenai performa model, terutama pada masalah klasifikasi dengan distribusi data yang tidak merata.